roc_jni/src/main/cpp/logger.cpp

```diff
@@ -65,23 +79,37 @@ void logger_handler(roc_log_level level, const char* component, const char* mess
         return;
     }
 
+    jint res;
+    int attached = 0; // know if detaching at the end is necessary
+    // check if it is needed to attach current thread
+    if ((res = handler_args.vm->GetEnv((void**) &env, JNI_VERSION)) == JNI_EDETACHED) {
 #ifdef __ANDROID__
-    handler_args.vm->AttachCurrentThread(&env, 0);
+        if (handler_args.vm->AttachCurrentThread(&env, 0) == JNI_OK)
```

Comment:
BTW, are we attaching/detaching Roc threads in every logger message? If so, how heavy is this operation? (I'd expect it's not very cheap).

If it's heavy, we could attach a thread once and register a thread destructor (e.g. [this way](https://lifecs.likai.org/2010/06/pthread-atexit.html)) to detach it. If this makes sense, could you please create an issue for that?

Reply:
> BTW, are we attaching/detaching Roc threads in every logger message? If so, how heavy is this operation? (I'd expect it's not very cheap).

If the current thread is already attached the `AttachCurrentThread` function will do no operation. However in the opposite case the answer is yes (a new java thread will be created as it can be saw to `attach_current_thread` function at [http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/tip/src/share/vm/prims/jni.cpp](http://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/tip/src/share/vm/prims/jni.cpp)).

Certainly we can call `DetachCurrentThread` before thread exits as it's mentioned here [https://developer.android.com/training/articles/perf-jni#threads](https://developer.android.com/training/articles/perf-jni#threads)

> you can use `pthread_key_create()` to define a destructor function that will be called before the thread exits, and call `DetachCurrentThread()` from there. (Use that key with `pthread_setspecific()` to store the `JNIEnv` in thread-local-storage).

I'll create an issue for that improvement feature :+1: